src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs

```diff
@@ -2732,6 +2732,121 @@ public bool Wait(int millisecondsTimeout, CancellationToken cancellationToken)
+        public Task WaitAsync(TimeSpan timeout, CancellationToken cancellationToken) =>
+            WaitAsync(ValidateTimeout(timeout, ExceptionArgument.timeout), cancellationToken);
+
+        private Task WaitAsync(uint millisecondsTimeout, CancellationToken cancellationToken)
+        {
+            if (IsCompleted || (!cancellationToken.CanBeCanceled && millisecondsTimeout == Timeout.UnsignedInfinite))
+            {
+                return this;
+            }
+
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return FromCanceled(cancellationToken);
+            }
+
+            if (millisecondsTimeout == 0)
+            {
+                return FromException(new TimeoutException());
+            }
+
+            return new CancellationPromise<VoidTaskResult>(this, millisecondsTimeout, cancellationToken);
+        }
+
+        /// <summary>Task that's completed when another task, timeout, or cancellation token triggers.</summary>
+        private protected sealed class CancellationPromise<TResult> : Task<TResult>, ITaskCompletionAction
+        {
+            /// <summary>The source task.  It's stored so that we can remove the continuation from it upon timeout or cancellation.</summary>
+            private readonly Task _task;
+            /// <summary>Cancellation registration used to unregister from the token source upon timeout or the task completing.</summary>
+            private readonly CancellationTokenRegistration _registration;
+            /// <summary>The timer used to implement the timeout.  It's stored so that it's rooted and so that we can dispose it upon cancellation or the task completing.</summary>
+            private readonly TimerQueueTimer? _timer;
+
+            internal CancellationPromise(Task source, uint millisecondsDelay, CancellationToken token)
+            {
+                Debug.Assert(source != null);
+                Debug.Assert(millisecondsDelay != 0);
+
+                // Register with the target task.
+                _task = source;
+                source.AddCompletionAction(this);
+
+                // Register with a timer if it's needed.
+                if (millisecondsDelay != Timeout.UnsignedInfinite)
+                {
+                    _timer = new TimerQueueTimer(static state =>
+                    {
+                        var thisRef = (CancellationPromise<TResult>)state!;
+                        if (thisRef.TrySetException(new TimeoutException()))
+                        {
+                            thisRef.Cleanup();
+                        }
+                    }, this, millisecondsDelay, Timeout.UnsignedInfinite, flowExecutionContext: false);
+                }
+
+                // Register with the cancellation token.
+                _registration = token.UnsafeRegister(static (state, cancellationToken) =>
```

Comment:
Is this a registration we could skip if `!token.CanBeCanceled`?

Reply:
All that saves is the call to UnsafeRegister: the first thing UnsafeRegister does is return a default registration if the token can't be canceled.